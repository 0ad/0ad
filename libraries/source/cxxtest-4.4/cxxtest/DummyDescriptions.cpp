/*
-------------------------------------------------------------------------
 CxxTest: A lightweight C++ unit testing library.
 Copyright (c) 2008 Sandia Corporation.
 This software is distributed under the LGPL License v3
 For more information, see the COPYING file in the top CxxTest directory.
 Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
 the U.S. Government retains certain rights in this software.
-------------------------------------------------------------------------
*/

#include <cxxtest/DummyDescriptions.h>

namespace CxxTest
{
DummyTestDescription::DummyTestDescription() {}

const char *DummyTestDescription::file() const { return "<no file>"; }
int DummyTestDescription::line() const { return 0; }
const char *DummyTestDescription::testName() const { return "<no test>"; }
const char *DummyTestDescription::suiteName() const { return "<no suite>"; }
bool DummyTestDescription::setUp() { return true;}
void DummyTestDescription::run() {}
bool DummyTestDescription::tearDown() { return true;}

TestDescription *DummyTestDescription::next() { return 0; }
const TestDescription *DummyTestDescription::next() const { return 0; }

DummySuiteDescription::DummySuiteDescription() : _test() {}

const char *DummySuiteDescription::file() const { return "<no file>"; }
int DummySuiteDescription::line() const { return 0; }
const char *DummySuiteDescription::suiteName() const { return "<no suite>"; }
TestSuite *DummySuiteDescription::suite() const { return 0; }
unsigned DummySuiteDescription::numTests() const { return 0; }
const TestDescription &DummySuiteDescription::testDescription(unsigned) const { return _test; }
SuiteDescription *DummySuiteDescription::next() { return 0; }
TestDescription *DummySuiteDescription::firstTest() { return 0; }
const SuiteDescription *DummySuiteDescription::next() const { return 0; }
const TestDescription *DummySuiteDescription::firstTest() const { return 0; }
void DummySuiteDescription::activateAllTests() {}
bool DummySuiteDescription::leaveOnly(const char * /*testName*/) { return false; }

bool DummySuiteDescription::setUp() { return true;}
bool DummySuiteDescription::tearDown() { return true;}

DummyWorldDescription::DummyWorldDescription() : _suite() {}

unsigned DummyWorldDescription::numSuites(void) const { return 0; }
unsigned DummyWorldDescription::numTotalTests(void) const { return 0; }
const SuiteDescription &DummyWorldDescription::suiteDescription(unsigned) const { return _suite; }
SuiteDescription *DummyWorldDescription::firstSuite() { return 0; }
const SuiteDescription *DummyWorldDescription::firstSuite() const { return 0; }
void DummyWorldDescription::activateAllTests() {}
bool DummyWorldDescription::leaveOnly(const char * /*suiteName*/, const char * /*testName*/) { return false; }

bool DummyWorldDescription::setUp() { return true;}
bool DummyWorldDescription::tearDown() { return true;}
}

