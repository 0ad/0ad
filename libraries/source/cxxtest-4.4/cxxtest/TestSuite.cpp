/*
-------------------------------------------------------------------------
 CxxTest: A lightweight C++ unit testing library.
 Copyright (c) 2008 Sandia Corporation.
 This software is distributed under the LGPL License v3
 For more information, see the COPYING file in the top CxxTest directory.
 Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
 the U.S. Government retains certain rights in this software.
-------------------------------------------------------------------------
*/

#ifndef __cxxtest__TestSuite_cpp__
#define __cxxtest__TestSuite_cpp__

#include <cxxtest/TestSuite.h>
#if defined(_CXXTEST_HAVE_STD)
#include <fstream>
#endif

namespace CxxTest
{
//
// TestSuite members
//
TestSuite::~TestSuite() {}
void TestSuite::setUp() {}
void TestSuite::tearDown() {}

//
// Test-aborting stuff
//
static bool currentAbortTestOnFail = false;

bool abortTestOnFail()
{
    return currentAbortTestOnFail;
}

void setAbortTestOnFail(bool value)
{
    currentAbortTestOnFail = value;
}

void doAbortTest()
{
#   if defined(_CXXTEST_HAVE_EH)
    if (currentAbortTestOnFail)
    {
        throw AbortTest();
    }
#   endif // _CXXTEST_HAVE_EH
}

//
// Max dump size
//
static unsigned currentMaxDumpSize = CXXTEST_MAX_DUMP_SIZE;

unsigned maxDumpSize()
{
    return currentMaxDumpSize;
}

void setMaxDumpSize(unsigned value)
{
    currentMaxDumpSize = value;
}

//
// Some non-template functions
//
void doTrace(const char *file, int line, const char *message)
{
    if (tracker().print_tracing)
    {
        tracker().trace(file, line, message);
    }
}

void doWarn(const char *file, int line, const char *message)
{
    tracker().warning(file, line, message);
}

#   if defined(_CXXTEST_HAVE_EH)
void doSkipTest(const char* file, int line, const char* message)
{
    tracker().skippedTest(file, line, message);
    throw SkipTest();
#   else
void doSkipTest(const char * file, int line, const char*)
{
    doWarn(file, line, "Test skipping is not supported without exception handling.");
#   endif
}

void doFailTest(const char * file, int line, const char * message)
{
    tracker().failedTest(file, line, message);
    TS_ABORT();
}

void doFailAssert(const char * file, int line,
                  const char * expression, const char * message)
{
    if (message)
    {
        tracker().failedTest(file, line, message);
    }
    tracker().failedAssert(file, line, expression);
    TS_ABORT();
}

bool sameData(const void * x, const void * y, unsigned size)
{
    if (size == 0)
    {
        return true;
    }

    if (x == y)
    {
        return true;
    }

    if (!x || !y)
    {
        return false;
    }

    const char *cx = (const char *)x;
    const char *cy = (const char *)y;
    while (size --)
    {
        if (*cx++ != *cy++)
        {
            return false;
        }
    }

    return true;
}

void doAssertSameData(const char * file, int line,
                      const char * xExpr, const void * x,
                      const char * yExpr, const void * y,
                      const char * sizeExpr, unsigned size,
                      const char * message)
{
    if (!sameData(x, y, size))
    {
        if (message)
        {
            tracker().failedTest(file, line, message);
        }
        tracker().failedAssertSameData(file, line, xExpr, yExpr, sizeExpr, x, y, size);
        TS_ABORT();
    }
}

#if defined(_CXXTEST_HAVE_STD)
bool sameFiles(const char * file1, const char * file2, std::ostringstream & explanation)
{
    std::string ppprev_line;
    std::string pprev_line;
    std::string prev_line;
    std::string curr_line;

    std::ifstream is1;
    is1.open(file1);
    std::ifstream is2;
    is2.open(file2);
    if (!is1)
    {
        explanation << "File '" << file1 << "' does not exist!";
        return false;
    }
    if (!is2)
    {
        explanation << "File '" << file2 << "' does not exist!";
        return false;
    }

    int nline = 1;
    char c1, c2;
    while (1)
    {
        is1.get(c1);
        is2.get(c2);
        if (!is1 && !is2) { return true; }
        if (!is1)
        {
            explanation << "File '" << file1 << "' ended before file '" << file2 << "' (line " << nline << ")";
            explanation << std::endl << "= " << ppprev_line << std::endl << "=  " << pprev_line << std::endl << "= " << prev_line << std::endl << "< " << curr_line;
            is1.get(c1);
            while (is1 && (c1 != '\n'))
            {
                explanation << c1;
                is1.get(c1);
            }
            explanation << std::endl;
            return false;
        }
        if (!is2)
        {
            explanation << "File '" << file2 << "' ended before file '" << file1 << "' (line " << nline << ")";
            explanation << std::endl << "= " << ppprev_line << std::endl << "=  " << pprev_line << std::endl << "= " << prev_line << std::endl << "> " << curr_line;
            is2.get(c2);
            while (is2 && (c2 != '\n'))
            {
                explanation << c2;
                is2.get(c2);
            }
            explanation << std::endl;
            return false;
        }
        if (c1 != c2)
        {
            explanation << "Files '" << file1 << "' and '" << file2 << "' differ at line " << nline;
            explanation << std::endl << "= " << ppprev_line << std::endl << "=  " << pprev_line << std::endl << "= " << prev_line;

            explanation << std::endl << "< " << curr_line;
            is2.get(c1);
            while (is1 && (c1 != '\n'))
            {
                explanation << c1;
                is2.get(c1);
            }
            explanation << std::endl;

            explanation << std::endl << "> " << curr_line;
            is2.get(c2);
            while (is2 && (c2 != '\n'))
            {
                explanation << c2;
                is2.get(c2);
            }
            explanation << std::endl;

            return false;
        }
        if (c1 == '\n')
        {
            ppprev_line = pprev_line;
            pprev_line = prev_line;
            prev_line = curr_line;
            curr_line = "";
            nline++;
        }
        else
        {
            curr_line += c1;
        }
    }
}
#endif

void doAssertSameFiles(const char * file, int line,
                       const char * file1, const char * file2,
                       const char * message)
{
#if defined(_CXXTEST_HAVE_STD)
    std::ostringstream explanation;
    if (!sameFiles(file1, file2, explanation))
    {
        if (message)
        {
            tracker().failedTest(file, line, message);
        }
        tracker().failedAssertSameFiles(file, line, file1, file2, explanation.str().c_str());
        TS_ABORT();
    }
#else
    tracker().failedAssertSameFiles(file, line, file1, file2, "This test is only supported when --have-std is enabled");
    TS_ABORT();
#endif
}

void doFailAssertThrows(const char * file, int line,
                        const char * expr, const char * type,
                        bool otherThrown,
                        const char * message,
                        const char * exception)
{
    if (exception)
    {
        tracker().failedTest(file, line, exception);
    }
    if (message)
    {
        tracker().failedTest(file, line, message);
    }

    tracker().failedAssertThrows(file, line, expr, type, otherThrown);
    TS_ABORT();
}

void doFailAssertThrowsNot(const char * file, int line,
                           const char * expression, const char * message,
                           const char * exception)
{
    if (exception)
    {
        tracker().failedTest(file, line, exception);
    }
    if (message)
    {
        tracker().failedTest(file, line, message);
    }

    tracker().failedAssertThrowsNot(file, line, expression);
    TS_ABORT();
}
}

#endif // __cxxtest__TestSuite_cpp__

