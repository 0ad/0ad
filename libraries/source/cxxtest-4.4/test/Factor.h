//
// This file is used to test WorldDescription::strTotalTests()
//

#include <cxxtest/TestSuite.h>
#include <cxxtest/DummyDescriptions.h>

class Factor : public CxxTest::TestSuite
{
public:
    class X : public CxxTest::DummyWorldDescription
    {
    public:
        unsigned n;
        unsigned numTotalTests() const { return n; }
    };

    X x;
    enum Limit { MAX_STRLEN_TOTAL_TESTS = CxxTest::WorldDescription::MAX_STRLEN_TOTAL_TESTS };
    char buffer[MAX_STRLEN_TOTAL_TESTS * 2];

    const char *convert(unsigned n)
    {
        x.n = n;
        return x.strTotalTests(buffer);
    }

    void test_Some_numbers()
    {
        TS_WARN(convert(53));
        for (unsigned n = 0; n < 64; ++ n)
        {
            TS_ASSERT_DIFFERS(n, 32);
            TS_WARN(convert(n));
        }
    }

    class ShorterThan
    {
    public:
        bool operator()(const char *s, unsigned n) const
        {
            unsigned len = 0;
            while (*s++ != '\0')
            {
                ++ len;
            }
            return (len < n);
        }
    };

    class NotShorterThan
    {
        ShorterThan _shorterThan;

    public:
        bool operator()(const char *s, unsigned n) const { return !_shorterThan(s, n); }
    };

    void test_Lengths()
    {
        unsigned reasonableLimit = 60060;
        for (unsigned n = 0; n < reasonableLimit; ++ n)
        {
            TS_ASSERT_RELATION(ShorterThan, convert(n), MAX_STRLEN_TOTAL_TESTS);
        }
        TS_ASSERT_RELATION(NotShorterThan, convert(reasonableLimit), MAX_STRLEN_TOTAL_TESTS);
    }
};
