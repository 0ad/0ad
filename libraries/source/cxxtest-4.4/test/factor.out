Running cxxtest tests (2 tests)
In Factor::test_Some_numbers:
Factor.h:30: Warning: 53 :)
Factor.h:33: Warning: 0
Factor.h:33: Warning: 1
Factor.h:33: Warning: 2 :|
Factor.h:33: Warning: 3 :|
Factor.h:33: Warning: 4 = 2^2
Factor.h:33: Warning: 5 :|
Factor.h:33: Warning: 6 = 2 * 3
Factor.h:33: Warning: 7 :|
Factor.h:33: Warning: 8 = 2^3
Factor.h:33: Warning: 9 = 3^2
Factor.h:33: Warning: 10 = 2 * 5
Factor.h:33: Warning: 11 :|
Factor.h:33: Warning: 12 = 2^2 * 3
Factor.h:33: Warning: 13 :|
Factor.h:33: Warning: 14 = 2 * 7
Factor.h:33: Warning: 15 = 3 * 5
Factor.h:33: Warning: 16 = 2^4
Factor.h:33: Warning: 17 :|
Factor.h:33: Warning: 18 = 2 * 3^2
Factor.h:33: Warning: 19 :|
Factor.h:33: Warning: 20 = 2^2 * 5
Factor.h:33: Warning: 21 = 3 * 7
Factor.h:33: Warning: 22 = 2 * 11
Factor.h:33: Warning: 23 :|
Factor.h:33: Warning: 24 = 2^3 * 3
Factor.h:33: Warning: 25 = 5^2
Factor.h:33: Warning: 26 = 2 * 13
Factor.h:33: Warning: 27 = 3^3
Factor.h:33: Warning: 28 = 2^2 * 7
Factor.h:33: Warning: 29 :|
Factor.h:33: Warning: 30 = 2 * 3 * 5
Factor.h:33: Warning: 31 :|
Factor.h:32: Error: Expected (n != 32), found (32)
Factor.h:33: Warning: 32 = 2^5
Factor.h:33: Warning: 33 = 3 * 11
Factor.h:33: Warning: 34 = 2 * 17
Factor.h:33: Warning: 35 = 5 * 7
Factor.h:33: Warning: 36 = 2^2 * 3^2
Factor.h:33: Warning: 37 :(
Factor.h:33: Warning: 38 = 2 * 19
Factor.h:33: Warning: 39 = 3 * 13
Factor.h:33: Warning: 40 = 2^3 * 5
Factor.h:33: Warning: 41 :(
Factor.h:33: Warning: 42 = 2 * 3 * 7
Factor.h:33: Warning: 43 :(
Factor.h:33: Warning: 44 = 2^2 * 11
Factor.h:33: Warning: 45 = 3^2 * 5
Factor.h:33: Warning: 46 = 2 * 23
Factor.h:33: Warning: 47 :(
Factor.h:33: Warning: 48 = 2^4 * 3
Factor.h:33: Warning: 49 = 7^2
Factor.h:33: Warning: 50 = 2 * 5^2
Factor.h:33: Warning: 51 = 3 * 17
Factor.h:33: Warning: 52 = 2^2 * 13
Factor.h:33: Warning: 53 :(
Factor.h:33: Warning: 54 = 2 * 3^3
Factor.h:33: Warning: 55 = 5 * 11
Factor.h:33: Warning: 56 = 2^3 * 7
Factor.h:33: Warning: 57 = 3 * 19
Factor.h:33: Warning: 58 = 2 * 29
Factor.h:33: Warning: 59 :(
Factor.h:33: Warning: 60 = 2^2 * 3 * 5
Factor.h:33: Warning: 61 :(
Factor.h:33: Warning: 62 = 2 * 31
Factor.h:33: Warning: 63 = 3^2 * 7
.
Failed 1 of 2 :( tests
Success rate: 50%
Error level = 1
